site/src/job_queue.rs

```diff
@@ -33,10 +34,97 @@ async fn enqueue_master_commits(ctxt: &Arc<SiteCtxt>) {
     }
 }
 
+// This function is split for testing purposes as mocking out `SiteCtxt` is non
+// trivial
+/// Given some pending requests and a list of completed requests determine if
+/// we have another request the we can process
+fn get_next_benchmark_request(
+    pending: &mut [BenchmarkRequest],
+    completed: &[BenchmarkRequest],
+) -> Option<BenchmarkRequest> {
+    // We now know we are only looking at commits which are possibly waiting
+    // for their parent to be complete.
+    pending.sort_by(|a, b| {
+        a.commit_type
+            .rank()
+            .cmp(&b.commit_type.rank())
+            .then_with(|| a.pr().unwrap_or(&0).cmp(b.pr().unwrap_or(&0)))
+            .then_with(|| a.created_at.cmp(&b.created_at))
+            .then_with(|| a.tag().cmp(b.tag()))
+    });
+
+    let completed_set: HashSet<String> = completed.iter().map(|r| r.tag().to_string()).collect();
+
+    let cutoff = Utc::now() - chrono::Duration::days(30);
+
+    let pending_by_sha: HashMap<String, &BenchmarkRequest> =
+        pending.iter().map(|r| (r.tag().to_string(), r)).collect();
+
+    for req in pending.iter() {
+        // release -> always ready
+        let parent_sha_opt = req.parent_sha();
+
+        let ready = match parent_sha_opt {
+            None => true,
+            Some(p_sha) => {
+                if completed_set.contains(p_sha) {
+                    true // parent is fresh & Completed
+                } else if let Some(parent_req) = pending_by_sha.get(p_sha) {
+                    parent_req.created_at < cutoff
+                } else {
+                    true // parent missing entirely
+                }
+            }
+        };
+
+        if ready {
+            return Some(req.clone());
+        }
+    }
+
+    None
+}
+
+/// Enqueue the job into the job_queue
+async fn enqueue_next_job(site_ctxt: &Arc<SiteCtxt>) {
+    let mut conn = site_ctxt.conn().await;
+    let mut pending = conn
+        .get_benchmark_requests_by_status(
+            &[
+                BenchmarkRequestStatus::InProgress,
+                BenchmarkRequestStatus::WaitingForParent,
+            ],
+            None,
+        )
+        .await;
+
+    // No requests to process or we have something currently in progress
+    if pending
+        .iter()
+        .any(|r| r.status == BenchmarkRequestStatus::InProgress)
+    {
+        return;
+    }
+
+    // We draw back the last 30 days of completed requests
+    let completed = conn
```

Comment:
This should be the list of *all* completed benchmark requests, that we ideally preload into a cached hashmap in memory to allow fast lookups (or something like that). The 30 day limit applies to the job queue, not to the benchmark requests. In theory, someone could do a try build with a very old parent (this is supported now), which was benchmarked more than 30 days ago.

Reply:
That's something a bit different, the old implementation caps last 30 days of master commits, not last 30 days of completed benchmark runs. As noted on the call, in theory someone could start a try build with e.g. a 2 month old master commit (it's rare, but it's something that could happen in the new design). So we should just load all completed requests from the DB. Right now let's load them from the DB, as the new requests table is still small, in the future we should cache them in memory.